src/cryptonote_core/blockchain_based_list.cpp

```diff
@@ -47,57 +48,25 @@ bool is_valid_stake(uint64_t block_height, uint64_t stake_block_height, uint64_t
   return true;
 }
 
-size_t extract_index(const char* it, size_t length)
-{
-  size_t result = 0;
-
-  for (;length--; it++)
-    result = (result << 8) + size_t(*it);
-
-  return result;
 }
 
-}
-
-void BlockchainBasedList::select_supernodes(const crypto::hash& block_hash, size_t items_count, const supernode_array& src_list, supernode_array& dst_list)
+void BlockchainBasedList::select_supernodes(size_t items_count, const supernode_array& src_list, supernode_array& dst_list)
 {
   size_t src_list_size = src_list.size();
 
   if (items_count > src_list_size)
     items_count = src_list_size;
 
-  static const size_t block_hash_items_count = sizeof(block_hash.data) / sizeof(*block_hash.data);
-
-  size_t step                 = size_t(log2(double(src_list.size())) / 8) + 1,
-         max_iterations_count = block_hash_items_count / step;   
-
-  if (items_count > max_iterations_count)
-    items_count = max_iterations_count;
-
-  decltype(&block_hash.data[0]) hash_ptr = &block_hash.data[0];
-
-  for (size_t i=0; i<items_count; i++, hash_ptr += step)
+  for (size_t i=0; i<src_list_size; i++)
   {
-    size_t base_index = extract_index(hash_ptr, step);
-
-    for (size_t offset=0;; offset++)
-    {    
-      if (offset == src_list_size)
-        return; //all supernodes have been selected
-
-      size_t               supernode_index     = (offset + base_index) % src_list_size;
-      const supernode& src_desc            = src_list[supernode_index];
-      const std::string&   supernode_public_id = src_desc.supernode_public_id;
-
-      if (selected_supernodes_cache.find(supernode_public_id) != selected_supernodes_cache.end())
-        continue; //supernode has been already selected
+    size_t random_value = m_rng() % (src_list_size - i);
 
-      dst_list.push_back(src_desc);
+    if (random_value >= items_count)
+      continue;
 
-      selected_supernodes_cache.insert(supernode_public_id);
+    dst_list.push_back(src_list[i]);
 
-      break;
-    }
+    items_count--;
   }
```

Comment:
The code above is going to maintain a sorted list, because elements are appended with random probabilities but in order.  This isn't necessarily a problem--the sample is still uniformly random--but if the later selection (inside `graft-ng`) isn't uniform then there will be a slight bias based on the registration height.  Two options for correcting this fairly easily:

Solution 1: make sure the sampling in graft-ng is uniform random.  I believe it still has a slight bias towards earlier elements in the blockchain list.

Solution 2: add a deterministic shuffle of the SN list here, before returning but after the sample has been selected.  This will most closely match the randomized order of the sampling that was being done here before switching to a Mersenne twister RNG:
```C++
/** Shuffle an array; essentially the same as std::shuffle, but deterministic given the same URNG sequence,
  * and with marginally inferior (but simpler) randomization using only `g() % M`.
  */
template <typename RandomIt, typename URNG>
void shuffle(RandomIt first, RandomIt last, URNG &&g) {
    for (auto i = last - first - 1; i > 0; --i) {
        using std::swap;
        swap(first[i], first[g() % (i+1)]);
    }
}



// ... at the end of select_supernodes:
    shuffle(dst_list.begin(), dst_list.end(), m_rng);
```